### PR TITLE
yul proto fuzzer: Add EVM version field

### LINKED ISSUE #7859: [yul] proto fuzzer: Add evm version field
## Abstract

The yul spec should include a field called `evm_version` (hard fork) so that program generation could be targeted at specific evm versions.

## Motivation

One way to do this is to add a protobuf field called evm version and map that to the EVMVersion class. We can then use functions from that class to decide whether to add a specific opcode or not.

### PR DESCRIPTION
**Pull request by bshastry**
Fixes #7859

### CODE REVIEW

File: test/tools/ossfuzz/protoToYul.cpp [hunk 1]
```diff
@@ -30,6 +30,7 @@
 using namespace std;
 using namespace solidity::yul::test::yul_fuzzer;
 using namespace solidity::yul::test;
+using namespace solidity::langutil;
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToYul.cpp:33]: Needed to make use of `EVMVersion` more convenient.

File: test/tools/ossfuzz/protoToYul.cpp [hunk 2]
```diff
@@ -86,6 +87,29 @@ string ProtoConverter::createAlphaNum(string const& _strBytes)
...
+		return EVMVersion::constantinople();
+	case Program::PETERSBURG:
+		return EVMVersion::petersburg();
+	case Program::ISTANBUL:
+		return EVMVersion::istanbul();
+	case Program::BERLIN:
+		return EVMVersion::berlin();
+	}
+}
+
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToYul.cpp:112]: This routine converts protobuf specification of EVM version to its respective `langutil::EVMVersion` representation

File: test/tools/ossfuzz/protoToYul.cpp [hunk 3]
```diff
@@ -230,7 +254,16 @@ void ProtoConverter::visit(Expression const& _x)
...
 {
-	switch (_x.op())
+	BinaryOp_BOp op = _x.op();
+
+	if ((op == BinaryOp::SHL || op == BinaryOp::SHR || op == BinaryOp::SAR) &&
+		!m_evmVersion.hasBitwiseShifting())
+	{
+		m_output << dictionaryToken();
+		return;
+	}
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToYul.cpp:264]: If we visit a binary operation that is one of shift (left/right) or arithmetic right shift, check whether the targeted evm version has these opcodes. If not, output a pseudo-random dictionary token.

File: test/tools/ossfuzz/protoToYul.cpp [hunk 4]
```diff
@@ -266,12 +299,15 @@ void ProtoConverter::visit(BinaryOp const& _x)
...
 	case BinaryOp::SHR:
+		yulAssert(m_evmVersion.hasBitwiseShifting(), "Proto fuzzer: Invalid evm version");
 		m_output << "shr";
 		break;
 	case BinaryOp::SHL:
+		yulAssert(m_evmVersion.hasBitwiseShifting(), "Proto fuzzer: Invalid evm version");
 		m_output << "shl";
 		break;
 	case BinaryOp::SAR:
+		yulAssert(m_evmVersion.hasBitwiseShifting(), "Proto fuzzer: Invalid evm version");
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToYul.cpp:310]: If we are down the binary operation conversion pipeline for the said shifts, assert that we are targeting a valid evm version

File: test/tools/ossfuzz/protoToYul.cpp [hunk 5]
```diff
@@ -475,7 +511,17 @@ void ProtoConverter::visit(TypedVarDecl const& _x)
...
-	switch (_x.op())
+	UnaryOp_UOp op = _x.op();
+
+	// Replace calls to extcodehash on unsupported EVMs with a dictionary
+	// token.
+	if (op == UnaryOp::EXTCODEHASH && !m_evmVersion.hasExtCodeHash())
+	{
+		m_output << dictionaryToken();
+		return;
+	}
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToYul.cpp:522]: Sanity check for extcodehash

File: test/tools/ossfuzz/protoToYul.cpp [hunk 6]
```diff
@@ -550,7 +596,12 @@ void ProtoConverter::visit(NullaryOp const& _x)
 		m_output << "codesize()";
 		break;
 	case NullaryOp::RETURNDATASIZE:
-		m_output << "returndatasize()";
+		// If evm supports returndatasize, we generate it. Otherwise,
+		// we output a dictionary token.
+		if (m_evmVersion.supportsReturndata())
+			m_output << "returndatasize()";
+		else
+			m_output << dictionaryToken();
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToYul.cpp:604]: Sanity check for returndatasize

File: test/tools/ossfuzz/protoToYul.cpp [hunk 7]
```diff
@@ -583,10 +634,20 @@ void ProtoConverter::visit(NullaryOp const& _x)
...
 		break;
 	case NullaryOp::CHAINID:
-		m_output << "chainid()";
+		// Replace calls to chainid() on unsupported EVMs with a dictionary
+		// token.
+		if (m_evmVersion.hasChainID())
+			m_output << "chainid()";
+		else
+			m_output << dictionaryToken();
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToYul.cpp:650]: Sanity checks for selfbalance and chainid

File: test/tools/ossfuzz/protoToYul.cpp [hunk 8]
```diff
@@ -609,6 +675,7 @@ void ProtoConverter::visit(CopyFunc const& _x)
 		m_output << "codecopy";
 		break;
 	case CopyFunc::RETURNDATA:
+		yulAssert(m_evmVersion.supportsReturndata(), "Proto fuzzer: Invalid evm version");
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToYul.cpp:678]: Sanity check for returndatacopy

File: test/tools/ossfuzz/protoToYul.cpp [hunk 9]
```diff
@@ -902,6 +979,7 @@ void ProtoConverter::visit(LowLevelCall const& _x)
 		m_output << "delegatecall(";
 		break;
 	case LowLevelCall::STATICCALL:
+		yulAssert(m_evmVersion.hasStaticCall(), "Proto fuzzer: Invalid evm version");
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToYul.cpp:982]: Sanity check for staticcall

File: test/tools/ossfuzz/protoToYul.cpp [hunk 10]
```diff
@@ -927,6 +1005,15 @@ void ProtoConverter::visit(LowLevelCall const& _x)
...
 	Create_Type type = _x.createty();
+
+	// Replace a call to create2 on unsupported EVMs with a dictionary
+	// token.
+	if (type == Create::CREATE2 && !m_evmVersion.hasCreate2())
+	{
+		m_output << dictionaryToken();
+		return;
+	}
+
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToYul.cpp:1016]: Sanity check for create2

File: test/tools/ossfuzz/protoToYul.cpp [hunk 11]
```diff
@@ -1737,6 +1824,9 @@ void ProtoConverter::visit(Program const& _x)
 	// Initialize input size
 	m_inputSize = _x.ByteSizeLong();
 
+	// Record EVM Version
+	m_evmVersion = evmVersionMapping(_x.ver());
```
Comment (bshastry) [inline test/tools/ossfuzz/protoToYul.cpp:1829]: Convert protobuf evm version into its langutil counterpart

File: test/tools/ossfuzz/yulProtoFuzzer.cpp [hunk 12]
```diff
@@ -51,7 +53,7 @@ DEFINE_PROTO_FUZZER(Program const& _input)
 
 	// AssemblyStack entry point
 	AssemblyStack stack(
-		langutil::EVMVersion(),
+		version,
```
Comment (bshastry) [inline test/tools/ossfuzz/yulProtoFuzzer.cpp:56]: Target the evm specified by the fuzzer

File: test/tools/ossfuzz/yulProto_diff_ossfuzz.cpp [hunk 13]
```diff
@@ -97,7 +99,7 @@ DEFINE_PROTO_FUZZER(Program const& _input)
 	termReason = yulFuzzerUtil::interpret(
 		os2,
 		stack.parserResult()->code,
-		EVMDialect::strictAssemblyForEVMObjects(langutil::EVMVersion::berlin()),
+		EVMDialect::strictAssemblyForEVMObjects(version),
```
Comment (bshastry) [inline test/tools/ossfuzz/yulProto_diff_ossfuzz.cpp:102]: Target the evm version specified by the fuzzer.

**Review by bshastry — COMMENTED** [review]
Self-annotated PR

File: test/tools/ossfuzz/protoToYul.cpp [hunk 14]
```diff
@@ -230,7 +254,16 @@ void ProtoConverter::visit(Expression const& _x)
 
 void ProtoConverter::visit(BinaryOp const& _x)
 {
-	switch (_x.op())
+	BinaryOp_BOp op = _x.op();
+
+	if ((op == BinaryOp::SHL || op == BinaryOp::SHR || op == BinaryOp::SAR) &&
```
Comment (erak) [inline test/tools/ossfuzz/protoToYul.cpp:259]: Maybe add a temporary variable `isShiftOperation` and then have the condition only in one line?

File: test/tools/ossfuzz/protoToYul.cpp (hunk 4 above)
Comment (erak) [reply test/tools/ossfuzz/protoToYul.cpp:310]: Shall we add this to the assertion message? Would that help future debugging?

File: test/tools/ossfuzz/protoToYul.h [hunk 15]
```diff
@@ -56,6 +58,12 @@ class ProtoConverter
 	ProtoConverter(ProtoConverter&&) = delete;
 	std::string programToString(Program const& _input);
 
+	/// Returns evm version
+	solidity::langutil::EVMVersion version()
```
Comment (erak) [inline test/tools/ossfuzz/protoToYul.h:62]: Could be one line.

**Comment by erak** [comment]
Only left some minor comments. Looks good!

**Review by Marenz — DISMISSED** [review]
Couldn't find any obvious problems... LGTM

**Pull request merged**